sys/unix/nleshared.cc

```diff
@@ -0,0 +1,798 @@
+
+#if defined(__linux__) && defined(__x86_64__)
+
+#include <elf.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <sys/mman.h>
+#include <dlfcn.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <string_view>
+#include <string>
+#include <cstdio>
+#include <stdexcept>
+#include <system_error>
+#include <vector>
+#include <cstddef>
+#include <cstring>
+#include <unordered_map>
+#include <mutex>
+#include <unordered_set>
+
+namespace nleshared {
+
+struct Fd {
+  int fd = -1;
+  Fd() = default;
+  Fd(int fd) noexcept : fd(fd) {}
+  Fd(const Fd&) = delete;
+  Fd(Fd&& n) noexcept {
+    fd = std::exchange(n.fd, -1);
+  }
+  Fd& operator=(const Fd&) = delete;
+  Fd& operator=(Fd&& n) noexcept {
+    std::swap(fd, n.fd);
+    return *this;
+  }
+  operator int() const noexcept {
+    return fd;
+  }
+  int release() noexcept {
+    return std::exchange(fd, -1);
+  }
+  ~Fd() {
+    if (fd != -1) {
+      ::close(fd);
+    }
+  }
+  size_t fileSize() const {
+    struct ::stat s;
+    if (::fstat(fd, &s)) {
+      throw std::system_error(errno, std::system_category(), "fstat");
+    }
+    return s.st_size;
+  }
+};
+
+struct Mmap {
+  std::byte* data_ = nullptr;
+  size_t size_ = 0;
+  Mmap() = default;
+  Mmap(std::byte* data, size_t size) noexcept : data_(data), size_(size) {}
+  Mmap(const Mmap&) = delete;
+  Mmap(Mmap&& n) noexcept {
+    data_ = std::exchange(n.data_, nullptr);
+    size_ = n.size_;
+  }
+  Mmap& operator=(const Mmap&) = delete;
+  Mmap& operator=(Mmap&& n) noexcept {
+    std::swap(data_, n.data_);
+    std::swap(size_, n.size_);
+    return *this;
+  }
+  ~Mmap() {
+    if (data_) {
+      ::munmap(data_, size_);
+    }
+  }
+  static Mmap make(const Fd& fd, size_t offset, size_t size, bool write, bool shared) {
+    errno = 0;
+    void* ptr = ::mmap(nullptr, size, write ? PROT_READ | PROT_WRITE : PROT_READ, shared ? MAP_SHARED : MAP_PRIVATE, (int)fd, offset);
+    if (!ptr || ptr == MAP_FAILED) {
+      throw std::system_error(errno, std::system_category(), "mmap");
+    }
+    return {(std::byte*)ptr, size};
+  }
+  static Mmap shared(const Fd& fd, size_t offset, size_t size, bool write) {
+    return make(fd, offset, size, write, true);
+  }
+  static Mmap private_(const Fd& fd, size_t offset, size_t size, bool write) {
+    return make(fd, offset, size, write, false);
+  }
+  operator bool() const noexcept {
+    return data_;
+  }
+  operator std::byte*() const noexcept {
```

Comment:
(Not so much a review comment more of a C++ question:) Just for my own curiosity: What does this do? I couldn't find it in https://en.cppreference.com/w/cpp/types/byte

Also would `constexpr` make sense for these functions here now that we're doing C++17?

Reply:
It's just a cast operator so it can be converted to a byte*. It's not actually used anywhere, though.

I don't think `constexpr` makes sense as there is no way to mmap anything at compile-time, so this type has no use in a constexpr context.

Reply:
That makes sense, thanks!